src/client/Client.cc

```diff
@@ -12665,7 +12656,29 @@ int Client::_rename(Inode *fromdir, const char *fromname, Inode *todir, const ch
     req->dentry_unless = CEPH_CAP_FILE_EXCL;
 
     InodeRef oldin, otherin;
-    res = _lookup(fromdir, fromname, 0, &oldin, perm);
+    Inode *fromdir_root = nullptr;
+    Inode *todir_root = nullptr;
+    int mask = 0;
+    bool quota_check = false;
+    if (fromdir != todir) {
+      fromdir_root =
+        fromdir->quota.is_enable() ? fromdir : get_quota_root(fromdir, perm);
+      todir_root =
+        todir->quota.is_enable() ? todir : get_quota_root(todir, perm);
+
+      if (todir_root->quota.is_enable() && fromdir_root != todir_root) {
+        quota_check = true;
+
+        if (oldde->inode && oldde->inode->is_dir()) {
+          // hack: take advantage of CEPH_STAT_RSTAT mask to force send lookup
+          // request to auth MDS to get latest rstat for source dir even if
+          // its dentry cache and inode caps are satisfied.
+          mask |= CEPH_STAT_RSTAT;
```

Comment:
We also need to do this for `todir`, right? What if we have the reverse situation where the target directory rstat information is out of date?

Reply:
change are done.